Add Gridfinity Refined thumbscrew option

andrewguy opened the pull request:
I've implemented the thumbscrew option from Gridfinity Refined (https://www.printables.com/model/413761-gridfinity-refined), enabling an optional screw hole in the middle of each base unit.

I did this for my own benefit, but I noticed there was an open issue for this (#195).

I'm very new to OpenSCAD (only been a couple of days), so I may not have the best approach here, but it seems to work.

I've pulled in Ryan Colyer's threads-scad library to handle this - https://github.com/rcolyer/threads-scad. Included as a top-level `threads.scad` file. Not sure if this is how you want to handle library dependencies?

These thumbscrew holes print well on a Bambu Lab A1 mini with Bambu PLA Basic, and the thumbscrew file provided by Gridfinity Refined work well (engages well, easy to insert and minimal slack).

![gridfinity_thumbscrew](https://github.com/user-attachments/assets/b0aa4f7d-e217-42b7-aed2-8e095fead736)

EmperorArthur commented:
We need to discuss how we want to handle libraries.

There are four major options I can think of:
* Require the user to manually have it.  -- Obviously bad idea.
* Include it in something like an "External" folder.  -- Library updates involve copying files.
* Use git submodules. -- Need to add extra instructions for people pulling down code.
* Do not allow 3rd party libraries.

If 3rd party libraries are allowed, we will need to make sure they are compatible license wise.

So far I have deliberately stayed away from libraries, but if we do decide to use them, then I suggest we use [BSOL](https://github.com/revarbat/BOSL) or [BSOL2](https://github.com/BelfrySCAD/BOSL2/).  Either would replace both this threading library, and a significant amount of helper code.


# Changes Needed
Regardless, there will need to be changes to this PR.  Even if it is just putting the library in a separate folder, and making a note of where to get the latest version.

EmperorArthur commented on threads.scad:
See my latest comment.  At the least this needs to be in it's own folder, and the Readme updated with where this was downloaded from.

EmperorArthur commented on gridfinity-rebuilt-utility.scad:
```diff
@@ -291,15 +294,29 @@ module block_base(hole_options, off=0, size=[BASE_SIZE, BASE_SIZE]) {
-                center=true
-            );
+            if (thumbscrew) {
+                ScrewHole(15, 5, position=[0, 0, 0], pitch=1.5) {
```
If you'll take a look at what the `ScrewHole` module does, you'll see it's a wrapper around `diference(){children() ScrewThread(...)}` with some translations that aren't used here.

Since this is already inside a `difference()` statement, there's no need for a wrapper, and it both duplicates code and doesn't fit with how holes are handled in this function.

I'd suggest putting the if statement and a raw `ScrewThread(...)` call immediately before or after the section with the comment "4 holes".

andrewguy replied:
Thanks, I've refactored this to just do a call to ScrewHole() as suggested. Works well.

EmperorArthur commented on gridfinity-rebuilt-utility.scad:
```diff
@@ -268,7 +271,7 @@ module gridfinityBase(gx, gy, length, dx, dy, hole_options=bundle_hole_options()
  * @param size [x, y] size of a single base.  Only set if deviating from the standard!
  */
-module block_base(hole_options, off=0, size=[BASE_SIZE, BASE_SIZE]) {
+module block_base(hole_options, off=0, size=[BASE_SIZE, BASE_SIZE], thumbscrew=false) {
```
Don't forget to add the additional parameter as a comment, and an assertion to ensure the value being passed in is a bool.

EmperorArthur commented on gridfinity-rebuilt-utility.scad:
```diff
@@ -202,16 +203,18 @@ module cut_move(x, y, w, h) {
         is_bool(final_cut) &&
-        is_bool(only_corners)
+        is_bool(only_corners) &&
+        is_bool(thumbscrew)
```
Good job here.

EmperorArthur commented on gridfinity-rebuilt-utility.scad:
```diff
@@ -202,16 +203,18 @@ module cut_move(x, y, w, h) {
 /**
  *@summary Create the base of a gridfinity bin, or use it for a custom object.
  * @param length X,Y size of a single Gridfinity base.
+ * @param thumbscrew Enable "gridfinity-refined" thumbscrew hole in the center of each base unit. This is a ISO Metric Profile, 15.0mm size, M15x1.5 designation.
```
Good job here.

EmperorArthur reviewed (changes requested):
I like the idea, but there are a few tweaks I'd like to see.

The big thing is we need to determine how libraries are handled.

andrewguy commented:
Thanks @EmperorArthur for the feedback.

I've refactored into a call to ScrewHole(), although this introduces a few more 'magic numbers' that come from the `threads.scad` code (i.e. `1.01*outer_diam + 1.25*tolerance`, presuming this introduces a bit of slack for threading the screw). Worth it I think.

I've moved the `threads.scad` library into `/external/threads-scad/`. In terms of licence, `threads.scad` is released under a [CC0-1.0 license](https://github.com/rcolyer/threads-scad#CC0-1.0-1-ov-file), which means it should be fine to include here.

In terms of handling external libraries, the only other option which would allow this pull request to go through would be to use Git submodules. I'd be concerned that submodules would lead to more trouble than they're worth (e.g. people need to use `--recurse-submodule` if doing a `git clone`, which often isn't obvious).

If the decision is made to not use external libraries, obviously this PR is a non-starter, which is fine.

EmperorArthur commented on gridfinity-rebuilt-utility.scad:
```diff
@@ -284,6 +292,13 @@ module block_base(hole_options, off=0, size=[BASE_SIZE, BASE_SIZE]) {
+    thumbscrew_height = 5;
+    thumbscrew_tolerance = 0.4;
+    thumbscrew_tooth_angle = 30;
+    thumbscrew_pitch = 1.5;
```
I'd prefer these in standard.scad, but will move them myself.

EmperorArthur reviewed (approved):
Personally I'm more of a fan of git submodules, but for something like this that doesn't change often I don't see a problem with it.  I also brought up the license as a future thing, not something to worry about right now.

Approving this, since my fix to #225 will have conflicts with this code, and I don't want to trash your hard work.

EmperorArthur commented:
@andrewguy Since I'm not a maintainer I can't edit this pull request, and I'm not allowed to manually merge it.

Could you please fix the whites-pace issue so the checks pass.  I'll merge this once that's done.  You could also squash your commits, or I can have the merge do it.

andrewguy commented:
Whitespace is fixed and all the pre-commit checks pass. If you could squash on merge that would be great - that's my normal strategy, and I don't want to break something by trying to squash an existing branch. :sweat_smile: 

Thanks for the review and being happy to integrate this. Appreciate it.
